File: supabase/functions/agents/tools.ts
import { tool } from '@openai/agents-core';
import { AgentContext } from '../_shared/chatkit/agents.ts';
import { retrieveWeather, normalizeUnit as normalizeTemperatureUnit, WeatherLookupError } from './weather.ts';
import { renderWeatherWidget, weatherWidgetCopyText } from './sample_widget.ts';

// Constants for theme switching
const SUPPORTED_COLOR_SCHEMES = new Set(['light', 'dark']);
export const CLIENT_THEME_TOOL_NAME = 'switch_theme';

/**
 * Normalize color scheme input to 'light' or 'dark'
 */
function normalizeColorScheme(value: string): string {
  const normalized = value.trim().toLowerCase();
  if (SUPPORTED_COLOR_SCHEMES.has(normalized)) {
    return normalized;
  }
  if (normalized.includes('dark')) {
    return 'dark';
  }
  if (normalized.includes('light')) {
    return 'light';
  }
  throw new Error("Theme must be either 'light' or 'dark'.");
}

/**
 * Switch the theme between light and dark mode.
 * This is a client tool that triggers a theme change in the frontend.
 *
 * Matches Python switch_theme implementation
 *
 * NOTE: Using JSON Schema directly instead of Zod because:
 * - Zod schemas require strict: true in the agents SDK
 * - The zodResponsesFunction helper may not be available in Deno/npm environment
 * - JSON Schema is universally supported by all model providers
 */
export const switchTheme = tool({
  name: 'switch_theme',
  description: 'Switch the chat interface between light and dark color schemes.',
  parameters: {
    type: 'object',
    properties: {
      theme: {
        type: 'string',
        description: 'The theme to switch to: "light" or "dark"',
        enum: ['light', 'dark'],
      },
    },
    required: ['theme'],
    additionalProperties: false,
  },
  strict: false, // Using JSON Schema directly, not Zod
  execute: ({ theme }: { theme: string }, ctx: { context?: AgentContext }) => {
    console.log('[agents] switch_theme tool called with theme:', theme);
    console.log('[agents] Context type:', typeof ctx, 'ctx:', ctx);
    try {
      const requested = normalizeColorScheme(theme);
      console.log('[agents] Normalized theme to:', requested);

      // The tool receives a RunContext that wraps our AgentContext in the 'context' property
      const agentContext = ctx.context as AgentContext;
      if (!agentContext) {
        console.error('[agents] No AgentContext found in RunContext');
        return null;
      }

      agentContext.client_tool_call = {
        name: CLIENT_THEME_TOOL_NAME,
        arguments: { theme: requested },
      };
      console.log('[agents] Set client_tool_call:', agentContext.client_tool_call);
      return { theme: requested };
    } catch (error) {
      console.error('[agents] Failed to switch theme:', error);
      return null;
    }
  },
});

/**
 * Get current weather and forecast for a location
 * Matches Python get_weather implementation
 */
export const getWeather = tool({
  name: 'get_weather',
  description: 'Look up the current weather and upcoming forecast for a location and render an interactive weather dashboard.',
  parameters: {
    type: 'object',
    properties: {
      location: {
        type: 'string',
        description: 'City name, address, or landmark to look up',
      },
      unit: {
        type: 'string',
        description: 'Temperature unit - "celsius" or "fahrenheit" (defaults to fahrenheit)',
        enum: ['celsius', 'fahrenheit'],
      },
    },
    required: ['location'],
    additionalProperties: false,
  },
  strict: false,
  execute: async ({ location, unit }: { location: string; unit?: string }, ctx: { context?: AgentContext }) => {
    console.log('[WeatherTool] tool invoked', { location, unit });
    try {
      const normalizedUnit = normalizeTemperatureUnit(unit);
      console.log('[WeatherTool] normalized unit:', normalizedUnit);

      const data = await retrieveWeather(location, normalizedUnit);
      console.log('[WeatherTool] lookup succeeded', {
        location: data.location,
        temperature: data.temperature,
        unit: data.temperature_unit,
      });

      const widget = renderWeatherWidget(data);
      const copyText = weatherWidgetCopyText(data);
      console.log('[WeatherTool] widget built');

      const agentContext = ctx.context as AgentContext;
      if (!agentContext) {
        console.error('[WeatherTool] No AgentContext found in RunContext');
        throw new Error('Weather data is currently unavailable for that location.');
      }

      console.log('[WeatherTool] streaming widget');
      await agentContext.stream_widget(widget, copyText);
      console.log('[WeatherTool] widget streamed');

      const observed = data.observation_time ? data.observation_time.toISOString() : null;

      return {
        location: data.location,
        unit: normalizedUnit,
        observed_at: observed,
      };
    } catch (error) {
      console.error('[WeatherTool] error:', error);
      if (error instanceof WeatherLookupError) {
        throw new Error(error.message);
      }
      throw new Error('Weather data is currently unavailable for that location.');
    }
  },
});

File: api/tools.py
from typing import Final, Literal, Any
import logging
from agents import function_tool, RunContextWrapper
from chatkit.agents import AgentContext, ClientToolCall
from .weather import retrieve_weather, normalize_unit as normalize_temperature_unit, WeatherLookupError
from .sample_widget import render_weather_widget, weather_widget_copy_text

logger = logging.getLogger(__name__)

# Constants for theme switching
SUPPORTED_COLOR_SCHEMES: Final[frozenset[str]] = frozenset({"light", "dark"})
CLIENT_THEME_TOOL_NAME: Final[str] = "switch_theme"

def _normalize_color_scheme(value: str) -> str:
    """Normalize color scheme input to 'light' or 'dark'."""
    normalized = str(value).strip().lower()
    if normalized in SUPPORTED_COLOR_SCHEMES:
        return normalized
    if "dark" in normalized:
        return "dark"
    if "light" in normalized:
        return "light"
    raise ValueError("Theme must be either 'light' or 'dark'.")

@function_tool(
    description_override="Switch the chat interface between light and dark color schemes."
)
async def switch_theme(
    ctx: RunContextWrapper[AgentContext],
    theme: str,
) -> dict[str, str] | None:
    """Switch the theme between light and dark mode.

    This is a client tool that triggers a theme change in the frontend.
    """
    try:
        requested = _normalize_color_scheme(theme)
        ctx.context.client_tool_call = ClientToolCall(
            name=CLIENT_THEME_TOOL_NAME,
            arguments={"theme": requested},
        )
        return {"theme": requested}
    except Exception:
        logger.exception("Failed to switch theme")
        return None

async def _needs_weather_approval(_ctx: RunContextWrapper[AgentContext], params: dict[str, Any], _call_id: str) -> bool:
    """Check if weather tool needs approval for Berkeley."""
    location = params.get("location", "")
    return "Berkeley" in location or "berkeley" in location.lower()

@function_tool(
    description_override="Look up the current weather and upcoming forecast for a location and render an interactive weather dashboard.",
    needs_approval=_needs_weather_approval,
)
async def get_weather(
    ctx: RunContextWrapper[AgentContext],
    location: str,
    unit: Literal["celsius", "fahrenheit"] | str | None = None,
) -> dict[str, str | None]:
    """Get current weather and forecast for a location.

    Args:
        location: City name, address, or landmark to look up
        unit: Temperature unit - 'celsius' or 'fahrenheit' (defaults to fahrenheit)

    Returns:
        Dictionary with location, unit, and observation time
    """
    print("[WeatherTool] tool invoked", {"location": location, "unit": unit})
    try:
        normalized_unit = normalize_temperature_unit(unit)
    except WeatherLookupError as exc:
        print("[WeatherTool] invalid unit", {"error": str(exc)})
        raise ValueError(str(exc)) from exc

    try:
        data = await retrieve_weather(location, normalized_unit)
    except WeatherLookupError as exc:
        print("[WeatherTool] lookup failed", {"error": str(exc)})
        raise ValueError(str(exc)) from exc

    print(
        "[WeatherTool] lookup succeeded",
        {
            "location": data.location,
            "temperature": data.temperature,
            "unit": data.temperature_unit,
        },
    )
    try:
        widget = render_weather_widget(data)
        copy_text = weather_widget_copy_text(data)
        payload: Any
        try:
            payload = widget.model_dump()
        except AttributeError:
            payload = widget
        print("[WeatherTool] widget payload", payload)
    except Exception as exc:
        print("[WeatherTool] widget build failed", {"error": str(exc)})
        raise ValueError("Weather data is currently unavailable for that location.") from exc

    print("[WeatherTool] streaming widget")
    try:
        await ctx.context.stream_widget(widget, copy_text=copy_text)
    except Exception as exc:
        print("[WeatherTool] widget stream failed", {"error": str(exc)})
        raise ValueError("Weather data is currently unavailable for that location.") from exc

    print("[WeatherTool] widget streamed")

    observed = data.observation_time.isoformat() if data.observation_time else None

    return {
        "location": data.location,
        "unit": normalized_unit,
        "observed_at": observed,
    }

